File: scripts/tba-engine/personal-taste-questions.mjs
import { attributeIdentity, canonical, LIKING_VALUES, sameSelection, unique } from './personal-taste-insights.mjs';

// 빈 필드에 가능한 응답을 하나씩 넣어 현재 규칙의 해석 변화를 비교한다.
// 분기에는 확률을 붙이지 않으며, 가상 응답이나 가상 근거를 원자료·결과에 추가하지 않는다.
function decisionStates(model) {
  const states = new Map();
  for (const candidate of model.candidates) {
    const minimum = candidate.conditions.length ? model.policy.minConditionMeals : model.policy.minMeals;
    if (candidate.distribution.mealCount >= minimum) states.set(`liking:${canonical(candidate.conditions)}`, candidate.direction ?? candidate.status);
  }
  for (const pattern of model.fitPatterns) {
    const minimum = pattern.conditions.length ? model.policy.minConditionMeals : model.policy.minMeals;
    if (pattern.distribution.mealCount >= minimum) states.set(`fit:${canonical(pattern.conditions)}`, pattern.repeatedValue ?? pattern.status);
  }
  for (const pattern of model.overallPatterns) {
    const minimum = pattern.conditions.length ? model.policy.minConditionMeals : model.policy.minMeals;
    const repeated = pattern.cells.filter(cell => cell.mealIDs.length >= minimum).map(cell => [cell.overallValue, cell.attributeValue]);
    if (repeated.length) states.set(`overall:${canonical(pattern.conditions)}`, canonical(repeated));
  }
  return states;
}

function hypotheticalRecords(records, question, source, answer) {
  if (question.intent === 'exploration') {
    const seed = records.find(row => row.attribute === question.attribute && (row.reference ?? null) === (question.reference ?? null));
    if (!seed) return null;
    const base = { ...seed, observationId: '__question_liking__', experienceId: '__question_experience__', mealId: '__question_meal__',
      kind: 'attribute_liking', scale: 'attribute-three-category-v1', value: answer, target: 'unspecified', phase: 'unspecified',
      dishKindIDs: [], selectionEvidence: null, sourceSpans: [], conditionSources: [], phrase: '', confirmationStatus: 'explicit_user_choice' };
    return [...records, base, { ...base, observationId: '__question_intensity__', kind: 'sensory_intensity',
      scale: 'expression-strength-v1', value: question.proposedCondition.value }];
  }
  if (!source) return null;
  const belongs = row => row.experienceId === source.experienceId && attributeIdentity(row) === attributeIdentity(source)
    && row.target === source.target && row.phase === source.phase && sameSelection(row, source);
  if (question.facet === 'target' || question.facet === 'phase') {
    return records.map(row => belongs(row) ? { ...row, [question.facet]: answer } : row);
  }
  const kind = question.facet === 'liking' ? 'attribute_liking' : 'sensory_intensity';
  const scale = question.facet === 'liking' ? 'attribute-three-category-v1' : 'expression-strength-v1';
  // 충돌한 강도와 미응답은 다르다. 가상 답변으로 기존 모순을 지우지 않는다.
  if (records.some(row => belongs(row) && row.kind === kind)) return null;
  return [...records, { ...source, observationId: '__question_answer__', kind, scale, value: answer,
    phrase: '', sourceSpans: [], conditionSources: [], selectionEvidence: source.selectionEvidence
      ? { ...source.selectionEvidence, facet: question.facet, responseValue: answer, labelValue: '' } : null }];
}

export function assessQuestion(question, records, project, baseline) {
  const relevant = records.filter(row => attributeIdentity(row) === canonical([question.attribute, question.reference ?? null]));
  const source = question.intent === 'exploration' ? null : relevant.find(row => question.evidenceIDs.includes(row.observationId)
    && row.target === question.target && row.phase === question.phase
    && (question.facet !== 'liking' || row.kind === 'sensory_presence'));
  const responseSourceID = source?.observationId ?? null;
  const comparedMealCount = unique(relevant.map(row => row.mealId)).length;
  let answers;
  if (question.intent === 'exploration' || question.facet === 'liking') answers = LIKING_VALUES;
  else if (question.facet === 'intensity') answers = ['weak', 'medium', 'strong'];
  else answers = unique(relevant.map(row => row[question.facet]).filter(value => value && value !== 'unspecified'));
  const states = [decisionStates(baseline)];
  let alternativeCount = 0;
  // 어느 분기에서도 최소 관찰 수에 이르지 못하면 일반적인 미응답 보완으로 둔다.
  const relevantRatings = unique(relevant.filter(row => row.kind === 'attribute_liking' || row.kind === 'preference_fit').map(row => row.mealId)).length;
  if (relevantRatings + 1 >= Math.min(baseline.policy.minMeals, baseline.policy.minConditionMeals)) {
    for (const answer of answers) {
      const hypothetical = hypotheticalRecords(records, question, source, answer);
      if (!hypothetical) continue;
      states.push(decisionStates(project(hypothetical))); alternativeCount++;
    }
  }
  const keys = unique(states.flatMap(state => [...state.keys()]));
  const changedInterpretationCount = keys.filter(key => unique(states.map(state => state.get(key) ?? 'not_supported')).length > 1).length;
  return { ...question, responseSourceID, impact: { changedInterpretationCount, alternativeCount, comparedMealCount,
    basis: changedInterpretationCount ? 'answer_can_change_interpretation' : 'fills_missing_information' } };
}

export function questionPrecedes(left, right) {
  if (left.impact.changedInterpretationCount !== right.impact.changedInterpretationCount) return left.impact.changedInterpretationCount > right.impact.changedInterpretationCount;
  if (left.score !== right.score) return left.score > right.score;
  if (left.mealIDs.length !== right.mealIDs.length) return left.mealIDs.length > right.mealIDs.length;
  return left.key < right.key;
}

File: ios/TasteBuddy/Domain/Models/PersonalTasteQuestions.swift
import Foundation

extension PersonalTasteModelBuilder {
    // 단일 속성의 모형에서 반복 근거가 있는 해석 상태만 비교한다.
    private static func decisionStates(policy: PersonalTastePolicy, candidates: [PersonalTasteCandidate], fitPatterns: [PersonalTasteFitPattern], overallPatterns: [PersonalTasteOverallPattern]) -> [String: String] {
        var states: [String: String] = [:]
        for candidate in candidates {
            let minimum = candidate.conditions.isEmpty ? policy.minMeals : policy.minConditionMeals
            if candidate.distribution.mealCount >= minimum {
                states["liking:" + canonical(candidate.conditions)] = candidate.direction ?? candidate.status
            }
        }
        for pattern in fitPatterns {
            let minimum = pattern.conditions.isEmpty ? policy.minMeals : policy.minConditionMeals
            if pattern.distribution.mealCount >= minimum {
                states["fit:" + canonical(pattern.conditions)] = pattern.repeatedValue ?? pattern.status
            }
        }
        for pattern in overallPatterns {
            let minimum = pattern.conditions.isEmpty ? policy.minMeals : policy.minConditionMeals
            let repeated = pattern.cells.filter { $0.mealIDs.count >= minimum }.map { [$0.overallValue, $0.attributeValue] }
            if !repeated.isEmpty { states["overall:" + canonical(pattern.conditions)] = json(repeated) }
        }
        return states
    }

    private static func hypotheticalRecords(_ records: [PersonalTasteModelRecord], question: Question, source: PersonalTasteModelRecord?, answer: String) -> [PersonalTasteModelRecord]? {
        if question.intent == "exploration" {
            guard var seed = records.first(where: { $0.attribute == question.attribute && $0.reference == question.reference }), let proposed = question.proposedCondition else { return nil }
            seed.observationId = "__question_liking__"; seed.experienceId = "__question_experience__"; seed.mealId = "__question_meal__"
            seed.kind = "attribute_liking"; seed.scale = "attribute-three-category-v1"; seed.value = .text(answer)
            seed.target = "unspecified"; seed.phase = "unspecified"; seed.dishKindIDs = []; seed.selectionEvidence = nil
            seed.sourceSpans = []; seed.conditionSources = []; seed.phrase = ""; seed.confirmationStatus = "explicit_user_choice"
            var intensity = seed; intensity.observationId = "__question_intensity__"
            intensity.kind = "sensory_intensity"; intensity.scale = "expression-strength-v1"; intensity.value = .text(proposed.value)
            return records + [seed, intensity]
        }
        guard let source else { return nil }
        func belongs(_ row: PersonalTasteModelRecord) -> Bool {
            row.experienceId == source.experienceId && insightAttributeKey(row) == insightAttributeKey(source)
                && row.target == source.target && row.phase == source.phase && sameSelection(row, source)
        }
        if question.facet == "target" || question.facet == "phase" {
            return records.map { row in
                guard belongs(row) else { return row }; var copy = row
                if question.facet == "target" { copy.target = answer } else { copy.phase = answer }; return copy
            }
        }
        let kind = question.facet == "liking" ? "attribute_liking" : "sensory_intensity"
        // 충돌한 기존 평가를 빈 응답처럼 덮어쓰지 않는다.
        guard !records.contains(where: { belongs($0) && $0.kind == kind }) else { return nil }
        var copy = source; copy.observationId = "__question_answer__"; copy.kind = kind
        copy.scale = question.facet == "liking" ? "attribute-three-category-v1" : "expression-strength-v1"
        copy.value = .text(answer); copy.phrase = ""; copy.sourceSpans = []; copy.conditionSources = []
        if let selection = source.selectionEvidence {
            copy.selectionEvidence = .init(selectionID: selection.selectionID, type: selection.type, catalogVersion: selection.catalogVersion, labelSnapshot: selection.labelSnapshot, facet: question.facet, labelValue: "", responseValue: answer, relatedBubbleID: selection.relatedBubbleID, relatedBubbleLabel: selection.relatedBubbleLabel, resolution: selection.resolution)
        }
        return records + [copy]
    }

    static func assessQuestionGroups(_ questions: [String: Question], included: [PersonalTasteModelRecord], userID: String, policy: PersonalTastePolicy, candidates: [PersonalTasteCandidate], fitPatterns: [PersonalTasteFitPattern], overallPatterns: [PersonalTasteOverallPattern]) -> [String: Question] {
        struct Baseline { let records: [PersonalTasteModelRecord]; let states: [String: String] }
        var baselines: [String: Baseline] = [:], result: [String: Question] = [:]
        for key in questions.keys.sorted() {
            var question = questions[key]!
            let attributeKey = json([question.attribute, nullable(question.reference)])
            if baselines[attributeKey] == nil {
                let attributeRecords = included.filter { insightAttributeKey($0) == attributeKey }
                let experiences = Set(attributeRecords.map(\.experienceId))
                let records = attributeRecords + included.filter { $0.kind == "overall_liking" && experiences.contains($0.experienceId) }
                baselines[attributeKey] = .init(records: records, states: decisionStates(
                    policy: policy,
                    candidates: candidates.filter { $0.attribute == question.attribute && $0.reference == question.reference },
                    fitPatterns: fitPatterns.filter { $0.attribute == question.attribute && $0.reference == question.reference },
                    overallPatterns: overallPatterns.filter { $0.attribute == question.attribute && $0.reference == question.reference }
                ))
            }
            let baseline = baselines[attributeKey]!
            let relevant = baseline.records.filter { insightAttributeKey($0) == attributeKey }
            let source = question.intent == "exploration" ? nil : relevant.first {
                question.evidenceIDs.contains($0.observationId) && $0.target == question.target && $0.phase == question.phase
                    && (question.facet != "liking" || $0.kind == "sensory_presence")
            }
            let answers: [String]
            if question.intent == "exploration" || question.facet == "liking" { answers = likingValues }
            else if question.facet == "intensity" { answers = ["weak", "medium", "strong"] }
            else { answers = unique(relevant.map { question.facet == "target" ? $0.target : $0.phase }.filter { $0 != "unspecified" }) }
            var states = [baseline.states], alternativeCount = 0
            let relevantRatingMeals = Set(relevant.filter { ["attribute_liking", "preference_fit"].contains($0.kind) }.map(\.mealId)).count
            if relevantRatingMeals + 1 >= min(policy.minMeals, policy.minConditionMeals) {
                for answer in answers {
                    guard let hypothetical = hypotheticalRecords(baseline.records, question: question, source: source, answer: answer) else { continue }
                    let model = buildRecords(records: hypothetical, userID: userID, policy: policy, assessQuestions: false, includeEvidence: false)
                    states.append(decisionStates(policy: model.policy, candidates: model.candidates, fitPatterns: model.fitPatterns, overallPatterns: model.overallPatterns))
                    alternativeCount += 1
                }
            }
            let stateKeys = Set(states.flatMap { $0.keys })
            let changed = stateKeys.filter { key in Set(states.map { $0[key] ?? "not_supported" }).count > 1 }.count
            question.responseSourceID = source?.observationId
            question.impact = .init(changedInterpretationCount: changed, alternativeCount: alternativeCount, comparedMealCount: Set(relevant.map(\.mealId)).count, basis: changed > 0 ? "answer_can_change_interpretation" : "fills_missing_information")
            result[key] = question
        }
        return result
    }
}
